### PR TITLE
bug: handle windows service shutdowns gracefully

### PR DESCRIPTION
**Pull request by muffins**
This pull request fixes a couple of bugs in the osquery Windows service handling routines

* A regression exists introduced in 3.2.0 causing the worker service to deadlock. This is due to  boost::asio::detail::win_threads not being shutdown correctly. @ExceptionalHandler identified this, thanks for the triage :) The fix here was to call boost::asio::detail::win_thread::set_terminate_threads(true), to ensure that the APC scheduler in boost terminates threads when the http::Client destructor is called on Windows.
* The watcher shutdown handler is run twice due to us not updating the value after execution. This shouldn't cause any issues as it simply stops the dispatcher services twice, however we're executing code that we don't need to be
* The watcher does not wait for the main thread of execution to complete before updating the Windows SCM with the STOPPED status for our service. This was causing the process to be exited before the Dispatcher had cleared out all references to worker threads, which was causing a sigabort to be thrown and the daemon to crash

This PR additionally introduces a new windows integration test to spawn the system service and test configurations. Folks should feel free to use this test for Windows system integration tests to ensure no regressions on the behavior of the system service.

This PR addresses #4235

Note, this PR has a follow-up issue, #4285, we need more tests around the thrashing of windows services. I have an integration test for this, but it needs more ironing out, and I don't want it to prevent this PR from being reviewed. I'll follow up this PR with a commit for the integration tests shortly!

### CODE REVIEW

**Comment by osqueryer** [comment]
:-1: The commit 75b6e41c2736416ee333b4b2139f00e0e4a129e6 (Job results: [6854](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6854/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 75b6e41c2736416ee333b4b2139f00e0e4a129e6 (Job results: [1309](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/1309/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 63aaa6dd09ebdca0524af916b8a59e5b34f18956 (Job results: [6856](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6856/)) failed one or more tests (**Linux**).

File: osquery/core/process.h [hunk 1]
```diff
@@ -41,6 +41,9 @@ namespace osquery {
 /// Constant for an invalid process
 const auto kInvalidPid = (PlatformPidType)-1;
 
+/// The saved thread ID for shutdown to short-circuit raising a signal.
+static std::thread::id kMainThreadId;
```
Comment (muffins) [inline osquery/core/process.h]: Ah. This needs to be `extern`, as we're intending to open a handle to it and wait on the thread to exit.

**Comment by osqueryer** [comment]
:-1: The commit 78521f9dc39286f9e22ac141daac506d59e9a412 (Job results: [3755](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3755/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 42a48fa19e39f6f208f7965c42d816f7d7e01f35 (Job results: [3757](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3757/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 42a48fa19e39f6f208f7965c42d816f7d7e01f35 (Job results: [3758](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3758/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 42a48fa19e39f6f208f7965c42d816f7d7e01f35 (Job results: [6862](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6862/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 42a48fa19e39f6f208f7965c42d816f7d7e01f35 (Job results: [3759](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3759/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 42a48fa19e39f6f208f7965c42d816f7d7e01f35 (Job results: [3760](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3760/)) failed one or more tests (**Windows**).

**Pull request merged**